Mathlib/RingTheory/MvPolynomial/NewtonIdentities.lean

```diff
@@ -0,0 +1,304 @@
+
+private theorem pair_map_mem_pairs (t : Finset σ × σ) (h : t ∈ pairs σ k) :
+    pairMap σ t ∈ pairs σ k := by
+  rw [pairs, mem_filter, pairMap] at *
+  simp only [mem_univ, true_and] at h
+  split_ifs with h1
+  · simp only [h1, implies_true, true_and, and_true] at h
+    simp only [mem_univ, true_and, card_erase_of_mem h1, tsub_le_iff_right, mem_erase, ne_eq, h1]
+    refine ⟨le_step h, ?_⟩
+    by_contra h2
+    rw [← h2] at h
+    exact not_le_of_lt (sub_lt (card_pos.mpr ⟨t.snd, h1⟩) zero_lt_one) h
+  · simp only [h1] at h
+    simp only [mem_univ, true_and, card_cons, mem_cons, true_or, implies_true, and_true]
+    exact (le_iff_eq_or_lt.mp h.left).resolve_left h.right
+
+@[simp] private theorem pair_map_involutive : (pairMap σ).Involutive := by
+  intro t
+  rw [pairMap, pairMap]
+  split_ifs with h1 h2 h3
+  · simp at h2
+  · simp [insert_erase h1]
+  · simp_all
+  · simp at h3
+
+private theorem weight_compose_pair_map (t : Finset σ × σ) (h : t ∈ pairs σ k) :
+    weight σ R k t + weight σ R k (pairMap σ t) = 0 := by
+  rw [pairMap, weight, weight]
+  rw [pairs, mem_filter] at h
+  have h2 (n : ℕ) : -(-1 : MvPolynomial σ R) ^ n = (-1) ^ (n + 1) := by
+    rw [← neg_one_mul ((-1 : MvPolynomial σ R) ^ n), pow_add, pow_one, mul_comm]
+  split_ifs with h1
+  · simp only [card_erase_of_mem h1, ← prod_erase_mul t.fst (fun j ↦ (X j : MvPolynomial σ R)) h1,
+      mul_comm, mul_assoc (∏ a in erase t.fst t.snd, X a), ← mul_add]
+    nth_rewrite 1 [← pow_one (X t.snd)]
+    simp only [← pow_add, add_comm]
+    have h3 : card t.fst ≥ 1 := lt_iff_add_one_le.mp (card_pos.mpr ⟨t.snd, h1⟩)
+    rw [← tsub_tsub_assoc h.right.left h3,
+      ← neg_neg ((-1 : MvPolynomial σ R) ^ (card t.fst - 1)), h2 (card t.fst - 1),
+      Nat.sub_add_cancel h3]
+    simp
+  · simp only [card_cons, prod_cons, mul_comm, mul_assoc (∏ a in t.fst, X a), ← mul_add]
+    nth_rewrite 2 [← pow_one (X t.snd)]
+    simp only [← pow_add,
+      ← Nat.add_sub_assoc (Nat.lt_of_le_of_ne h.right.left (mt h.right.right h1)), add_comm]
+    rw [← neg_neg ((-1 : MvPolynomial σ R) ^ card t.fst), h2]
+    simp
+
+private theorem weight_zero_for_fixed_by_pair_map (t : Finset σ × σ) (h : t ∈ pairs σ k)
+    (h2 : pairMap σ t = t) : weight σ R k t = 0 := by
+  have h3 := weight_compose_pair_map σ R t h
+  rw [h2, ← two_mul, _root_.mul_eq_zero] at h3
+  exact h3.resolve_left two_ne_zero
+
+private theorem weight_zero_for_fixed_by_pair_map' (t : Finset σ × σ) (h : t ∈ pairs σ k)
+    (h1 : weight σ R k t ≠ 0) : pairMap σ t ≠ t := mt (weight_zero_for_fixed_by_pair_map σ R t h) h1
+
+private theorem weight_sum (k : ℕ) : ∑ t in pairs σ k, weight σ R k t = 0 :=
+  sum_involution (fun t _ ↦ pairMap σ t) (weight_compose_pair_map σ R)
+    (weight_zero_for_fixed_by_pair_map' σ R) (pair_map_mem_pairs σ)
+    (fun t _ ↦ pair_map_involutive σ t)
+
+private theorem sum_equiv_k (k : ℕ) (f : Finset σ × σ → MvPolynomial σ R) :
```

Comment:
Again, trying to name this based on type would yield something like:
```suggestion
private theorem sum_filter_pairs_eq_sum_powersetLen_sum
    (k : ℕ) (f : Finset σ × σ → MvPolynomial σ R) :
```

Reply:
Done!